qiskit_experiments/framework/json.py

```diff
@@ -0,0 +1,516 @@
+            settings = {
+                "data": obj.data,
+                "input_dims": obj.input_dims(),
+                "output_dims": obj.output_dims(),
+            }
+            return _serialize_object(obj, settings=settings)
+        if isinstance(obj, DensityMatrix):
+            # Temporary fix for incorrect settings in qiskit-terra
+            # See https://github.com/Qiskit/qiskit-terra/pull/7194
+            settings = {
+                "data": obj.data,
+                "dims": obj.dims(),
+            }
+            return _serialize_object(obj, settings=settings)
+        if is_type(obj):
+            return _serialize_type(obj)
+        try:
+            return super().default(obj)
+        except TypeError:
+            return _serialize_object(obj)
+
+
+class ExperimentDecoder(json.JSONDecoder):
+    """JSON Decoder for Qiskit Experiments.
+
+    This class extends the default Python JSONDecoder by including built-in
+    support for all objects that that can be serialized using the
+    :class:`ExperimentEncoder`.
+
+    See :class:`ExperimentEncoder` class documentation for details.
+    """
+
+    _NaNs = {"NaN": math.nan, "Infinity": math.inf, "-Infinity": -math.inf}
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(object_hook=self.object_hook, *args, **kwargs)
+
+    def object_hook(self, obj):
+        """Object hook."""
+        if "__type__" in obj:
+            obj_type = obj["__type__"]
+            obj_val = obj["__value__"]
+            if obj_type == "complex":
+                return obj_val[0] + 1j * obj_val[1]
+            if obj_type == "ndarray":
+                return _decode_and_deserialize(obj_val, np.load, name=obj_type)
+            if obj_type == "spmatrix":
+                return _decode_and_deserialize(obj_val, sps.load_npz, name=obj_type)
+            if obj_type == "b64encoded":
+                return _deserialize_bytes(obj_val)
+            if obj_type == "set":
+                return set(obj_val)
+            if obj_type == "QuantumCircuit":
+                return _decode_and_deserialize(obj_val, qpy_serialization.load, name=obj_type)[0]
+            if obj_type == "ParameterExpression":
+                return _decode_and_deserialize(
+                    obj_val, qpy_serialization._read_parameter_expression, name=obj_type
+                )
+            if obj_type == "Result":
+                return Result.from_dict(obj_val)
+            if obj_type == "safe_float":
+                return self._NaNs.get(obj_val, obj_val)
+            if obj_type == "object":
+                return _deserialize_object(obj_val)
+            if obj_type == "type":
+                return _deserialize_type(obj_val)
+
+            # Deprecated formats
+            if obj_type == "array":
+                _deprecation_warning(obj_type, "0.3.0")
+                return np.array(obj_val)
+            if obj_type == "function":
```

Comment:
We cannot directly set callback to experiment configuration?

Reply:
This is a bit beyond JSON serialization and I'm not sure is a good idea in this case. You also won't be able to load dynamically generated functions, which callbacks often are since they are wrappers for other functions.